### PR TITLE
Add JsonDocCk Tool for rustdoc-json

### PR DESCRIPTION
**Pull request by CraftSpider**
Implements a new test system for rustdoc JSON output, jsondocck. Modeled after htmldocck, this tool reads directives in the test file and checks them against the output. These directives use JSONPath, a pair to XPath for json. This obsoletes the old strict subset tool, allowing both finer-grained control of what is tested and better errors on failure.

Not sure on the changes to Cargo.lock, I can back that out if needed.

r? @jyn514

### CODE REVIEW

File: src/tools/jsondocck/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+                        lineno,
+                    );
+                    return false;
+                }
+                if args[2].parse::<usize>().is_err() {
+                    print_err(&format!("Third argument to @count must be a valid usize"), lineno);
+                    return false;
+                }
+            }
+        }
```
Comment (jyn514) [inline src/tools/jsondocck/src/main.rs:88]:
```suggestion
        if args[0] == "-" && command_num == 0 {
            print_err(
                &format!("Tried to use the previous path in the first command"),
                lineno,
            );
            return false;
        }
        if let Command::Count = self {
            if args[2].parse::<usize>().is_err() {
                print_err(&format!("Third argument to @count must be a valid usize"), lineno);
                return false;
            }
        }
```

File: src/tools/jsondocck/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+        let cap = match LINE_PATTERN.captures(line) {
+            Some(c) => c,
+            None => continue,
+        };
+
+        let negated = match cap.name("negated") {
+            Some(m) => m.as_str() == "!",
+            None => false,
+        };
```
Comment (jyn514) [inline src/tools/jsondocck/src/main.rs:138]: This should never be `None`, right? Because `negated` is hard-coded in the regex?
```suggestion
        let negated = cap.name("negated").unwrap().as_str() == "!";
```

File: src/tools/jsondocck/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+                            let pat: Value = serde_json::from_str(&command.args[2]).unwrap();
+
+                            !results.is_empty() && results.into_iter().any(|val| *val == pat)
+                        }
+                        Err(_) => false,
+                    }
+                }
+                _ => {
+                    unreachable!()
+                }
```
Comment (jyn514) [inline src/tools/jsondocck/src/main.rs:214]:
```suggestion
                _ => unreachable!(),
```

File: src/tools/jsondocck/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+            }
+        }
+        CommandKind::Count => {
+            match command.args.len() {
+                // @count <path> <jsonpath> <count> = Check that the jsonpath matches exactly [count] times
+                3 => {
```
Comment (jyn514) [inline src/tools/jsondocck/src/main.rs:220]:
```suggestion
       assert_eq!(command.args.len(), 3);
```
and then you can deindent everything following.

**Review by jyn514 — COMMENTED** [review]
This looks great :) Can you add a test that intentionally fails and post the error output?

File: src/tools/jsondocck/src/cache.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        if path != "-" {
+            let resolve = self.root.join(path);
+            self.last_path = Some(resolve.clone());
+            Ok(resolve)
+        } else {
+            match &self.last_path {
+                Some(p) => Ok(p.clone()),
+                None => unreachable!(),
+            }
+        }
```
[the remark this answers is not in the document]
Comment (CraftSpider) [reply src/tools/jsondocck/src/cache.rs:35]: Ah, thanks, nice catch! Leftover from when I first wrote it, hadn't yet moved all 'validity' checks into the collection step

File: src/tools/jsondocck/src/main.rs (hunk 1 above)
Comment (CraftSpider) [reply src/tools/jsondocck/src/main.rs:88]: I considered this, and wasn't sure about if we ever added commands that didn't start with a filename. On the other hand, it's not hard to change, so I'll swap it.

Comment (jyn514) [reply src/tools/jsondocck/src/main.rs:88]: I would rather make the existing code easier to read (and less duplicated) than try to account for unknown changes in the future I think.

File: src/tools/compiletest/src/runtest.rs [hunk 6]
```diff
@@ -2490,28 +2490,43 @@ impl<'test> TestCx<'test> {
...
+        //     Command::new(&self.config.docck_python)
+        //         .arg(root.join("src/test/rustdoc-json/compare.py"))
+        //         .arg(&expected)
+        //         .arg(&json_out)
+        //         .arg(&expected.parent().unwrap()),
+        // );
+        //
+        // if !res.status.success() {
+        //     self.fatal_proc_rec("compare failed!", &res);
+        // }
```
[the remark this answers is not in the document]
Comment (CraftSpider) [reply src/tools/compiletest/src/runtest.rs]: > This looks great :) Can you add a test that intentionally fails and post the error output?

Do you mean add a test to git, or just write one up and post the results?

**Comment by CraftSpider** [comment]
> This looks great :) Can you add a test that intentionally fails and post the error output?

Do you mean add one to git, or just write one up and post the results?

**Comment by jyn514** [comment]
Just write one up and post the results.

**Comment by CraftSpider** [comment]
<details><summary><strong>Failure Output</strong></summary>
<p>

```
running 8 tests
iiiiiiFF
failures:

---- [rustdoc-json] rustdoc-json\bad.rs stdout ----

error: jsondocck failed!
status: exit code: 1
command: "D:\\Dev\\Git\\rust\\build\\x86_64-pc-windows-msvc\\stage1-tools-bin\\j
sondocck.exe" "--doc-dir" "D:\\Dev\\Git\\rust\\build\\x86_64-pc-windows-msvc\\te
st\\rustdoc-json\\bad" "--template" "D:\\Dev\\Git\\rust\\src/test\\rustdoc-json\
\bad.rs"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
Invalid command: Tried to use the previous path in the first command on line 2
Invalid command: `!@has`, (help: try with `@!has`) on line 3
Invalid command: Incorrect number of arguments to `@has` on line 4
Invalid command: Unrecognized command name `@unknown` on line 5
Invalid command: Third argument to @count must be a valid usize on line 6
Error: "Jsondocck failed for D:\\Dev\\Git\\rust\\src/test\\rustdoc-json\\bad.rs"

------------------------------------------


---- [rustdoc-json] rustdoc-json\bad2.rs stdout ----

error: jsondocck failed!
status: exit code: 1
command: "D:\\Dev\\Git\\rust\\build\\x86_64-pc-windows-msvc\\stage1-tools-bin\\j
sondocck.exe" "--doc-dir" "D:\\Dev\\Git\\rust\\build\\x86_64-pc-windows-msvc\\te
st\\rustdoc-json\\bad2" "--template" "D:\\Dev\\Git\\rust\\src/test\\rustdoc-json
\\bad2.rs"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
Failed check: `@has bad2.json $.not.a.real.path` didn't match when it should on
line 2
Failed check: `@count bad2.json $.index.* 1` didn't match when it should on line
 3
Error: "Jsondocck failed for D:\\Dev\\Git\\rust\\src/test\\rustdoc-json\\bad2.rs
"

------------------------------------------



failures:
    [rustdoc-json] rustdoc-json\bad.rs
    [rustdoc-json] rustdoc-json\bad2.rs

test result: FAILED. 0 passed; 2 failed; 6 ignored; 0 measured; 0 filtered out;
finished in 6.99s
```
</p>
</details>

Note that the system won't display syntax errors at the same time as check errors. I could change this, it was just a bit easier to implement this way.

File: src/etc/check_missing_items.py [hunk 7]
```diff
@@ -4,6 +4,8 @@
 # `index` or `paths`. It DOES NOT check that the structure of the produced json is actually in
 # any way correct, for example an empty map would pass.
 
+# FIXME: Better error output
```
Comment (jyn514) [inline src/etc/check_missing_items.py:7]: Why did you move the file?

File: src/bootstrap/builder.rs [hunk 8]
```diff
@@ -751,6 +751,11 @@ impl<'a> Builder<'a> {
         cmd
     }
 
+    /// Gets a path to the jsondocck tool
+    pub fn jsondocck(&self, compiler: Compiler, target: TargetSelection) -> PathBuf {
+        self.ensure(tool::JsonDocCk { compiler, target })
+    }
+
```
Comment (jyn514) [inline src/bootstrap/builder.rs:758]: This is small enough I would rather just call `self.ensure` inline, that would also make it more clear it builds `JsonDocCk` if necessary.

File: src/bootstrap/tool.rs [hunk 9]
```diff
@@ -367,6 +367,7 @@ bootstrap_tool!(
     RustdocTheme, "src/tools/rustdoc-themes", "rustdoc-themes";
     ExpandYamlAnchors, "src/tools/expand-yaml-anchors", "expand-yaml-anchors";
     LintDocs, "src/tools/lint-docs", "lint-docs";
+    JsonDocCk, "src/tools/jsondocck", "jsondocck", is_unstable_tool = true;
```
Comment (jyn514) [inline src/bootstrap/tool.rs:370]: `is_unstable` is used to tell if `RUSTC_BOOTSTRAP` should be passed. I don't think there's any reason to use RUSTC_BOOTSTRAP here.
```suggestion
    JsonDocCk, "src/tools/jsondocck", "jsondocck";
```
(I'll see about removing this from compile test soon, then we could get rid of the option altogether)

File: src/test/rustdoc-json/nested.rs [hunk 10]
```diff
@@ -1,7 +1,25 @@
 // edition:2018
+
+// @has nested.json "$.index.['0:0'].kind" \"module\"
+// @has - "$.index.['0:0'].inner.is_crate" true
+// @has - "$.index.['0:0'].inner.items[*]" \"0:3\"
+
+// @has nested.json "$.index.['0:3'].kind" \"module\"
+// @has - "$.index.['0:3'].inner.is_crate" false
+// @has - "$.index.['0:3'].inner.items[*]" \"0:4\"
+// @has - "$.index.['0:3'].inner.items[*]" \"0:7\"
```
Comment (jyn514) [inline src/test/rustdoc-json/nested.rs:10]: This makes me very nervous. If the compiler changes how DefIds are generated, this will break.

Can you instead find `l1` by name, and just assert that the ID exists, not the format?

File: src/tools/compiletest/src/main.rs [hunk 11]
```diff
@@ -196,6 +197,7 @@ pub fn parse_config(args: Vec<String>) -> Config {
     let has_tidy = Command::new("tidy")
         .arg("--version")
         .stdout(Stdio::null())
+        .stderr(Stdio::null())
```
Comment (jyn514) [inline src/tools/compiletest/src/main.rs:200]: Does tidy write to stderr on your machine?

File: src/tools/jsondocck/src/config.rs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        std::process::exit(1);
+    }
+
+    let matches = opts.parse(args_).unwrap();
+
+    if matches.opt_present("h") || matches.opt_present("help") {
+        let message = format!("Usage: {} <doc-dir> <template>", argv0);
+        println!("{}", opts.usage(&message));
+        std::process::exit(1);
+    }
```
Comment (jyn514) [inline src/tools/jsondocck/src/config.rs:30]: Why do you check for `-h` twice?

File: src/tools/jsondocck/src/cache.rs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+        let file = fs::read_to_string(&path)?;
+
+        self.files.insert(path, file.clone());
+
+        Ok(file)
+    }
+
+    pub fn get_value(&mut self, path: &String) -> Result<Value, CkError> {
```
Comment (jyn514) [inline src/tools/jsondocck/src/cache.rs:49]:
```suggestion
    /// Parse the JSON from a file. If called multiple times, the file will only be read once.
    pub fn get_value(&mut self, path: &String) -> Result<Value, CkError> {
```

File: src/tools/jsondocck/src/cache.rs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        let path = self.resolve_path(path);
+
+        if let Some(v) = self.values.get(&path) {
+            return Ok(v.clone());
+        }
+
+        let file = fs::File::open(&path)?;
+
+        let val = serde_json::from_reader::<_, Value>(file)?;
```
Comment (jyn514) [inline src/tools/jsondocck/src/cache.rs:58]: Shouldn't this go through `self.get_file` so you don't have two separate caches?

File: src/tools/jsondocck/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+
+fn check_command(command: Command, cache: &mut Cache) -> Result<(), CkError> {
+    let result = match command.kind {
+        CommandKind::Has => {
+            match command.args.len() {
+                // @has <path> = file existence
+                1 => match cache.get_file(&command.args[0]) {
+                    Ok(_) => true,
+                    Err(_) => false,
+                },
```
Comment (jyn514) [inline src/tools/jsondocck/src/main.rs:174]:
```suggestion
                1 => cache.get_file(&command.args[0]).is_ok(),
```

File: src/bootstrap/tool.rs (hunk 9 above)
Comment (jyn514) [reply src/bootstrap/tool.rs:370]: Hmm, this was adding in 51b7044347d2cb9e2eae18ad49c79803dda60be2, which IIRC was related to something with bootstrapping libtest: https://github.com/rust-lang/rust/pull/68019. And it turns out that is actually necessary because the policy is to build against master libstd when using RUSTC_BOOTSTRAP for anything other than the compiler: https://github.com/rust-lang/rust/issues/59264

@cuviper this confused me at first because building with `Mode::ToolBootstrap` actually still allows using unstable features. Would it be possible to add `RUSTC_BOOTSTRAP` only when building in mode `ToolStd`?

Comment (jyn514) [reply src/bootstrap/tool.rs:370]: See https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Removing.20unnecessary.20RUSTC_BOOTSTRAP, this ended up harder to change than I expected.

Comment (cuviper) [reply src/bootstrap/tool.rs:370]: In principle, yes I agree that those would preferably not be allowed to use unstable features at all. In practice, that may be hard to isolate... as you found.

File: src/etc/check_missing_items.py (hunk 7 above)
Comment (CraftSpider) [reply src/etc/check_missing_items.py:7]: Other test helpers were in this folder, it seems the more 'standard' place. I can move it back or split it to a different PR, as it's somewhat unrelated.

File: src/bootstrap/builder.rs (hunk 8 above)
Comment (CraftSpider) [reply src/bootstrap/builder.rs:758]: Hm. I'd be alright with that, I added this for 'conformity', there are a bunch of these functions for things like 'rustc' or 'rustdoc'. On the other hand, those things are ensured in more than one place :P

File: src/tools/compiletest/src/main.rs (hunk 11 above)
Comment (CraftSpider) [reply src/tools/compiletest/src/main.rs:200]: Yes, and in fact it should be on everyone's machine. The first 3 lines of `tidy/src/main.rs`:
```rust
let root_path: PathBuf = env::args_os().nth(1).expect("need path to root of repo").into();
    let cargo: PathBuf = env::args_os().nth(2).expect("need path to cargo").into();
    let output_directory: PathBuf =
        env::args_os().nth(3).expect("need path to output directory").into();
```

Thinking about it, this check is actually broken and always false, because `--version` is entirely ignored and tidy just unconditionally panics. This is another change I can back out here and make into its own PR though, as it is unrelated to the function of the json tests.

File: src/tools/jsondocck/src/config.rs (hunk 12 above)
Comment (CraftSpider) [reply src/tools/jsondocck/src/config.rs:30]: I just copied this code from compiletest. I think it's so that invalid args are ignored, if the first one is `-h`. I can change it if you want.

File: src/tools/jsondocck/src/cache.rs (hunk 13 above)
Comment (CraftSpider) [reply src/tools/jsondocck/src/cache.rs:49]: :+1:, do you want me to do a general documentation pass?

File: src/tools/jsondocck/src/cache.rs (hunk 14 above)
Comment (CraftSpider) [reply src/tools/jsondocck/src/cache.rs:58]: Probably. This exactly mirrors the behavior of htmldocck for this function, I didn't think too hard about why it did it that way. I'll swap it over.

File: src/test/rustdoc-json/nested.rs (hunk 10 above)
Comment (CraftSpider) [reply src/test/rustdoc-json/nested.rs:10]: Do you mean not using the `.index.['n:m']` part, or the `items[*] "n:m"` part? The first one I could do by using filter expressions, though it would make the paths a little more verbose. The second I actually think it might be better to use `@count` and just say 'there are two things in the items list', as I can't think of a good way to assert they're the right items aside from hardcoding the IDs

File: src/tools/compiletest/src/main.rs (hunk 11 above)
Comment (jyn514) [reply src/tools/compiletest/src/main.rs:200]: @CraftSpider there are two different tidy commands. One is `src/tools/tidy`, one is https://www.w3.org/People/Raggett/tidy/. This is checking HTML tidy.

Can you revert this change?

Comment (CraftSpider) [reply src/tools/compiletest/src/main.rs:200]: Interesting. It was definitely invoking the one I pointed to, I was getting the full panic. I'll revert the change and look into why.

File: src/etc/check_missing_items.py (hunk 7 above)
Comment (jyn514) [reply src/etc/check_missing_items.py:7]: Hmm, given that `htmldocck` is here it seems fine to keep check_missing_items here too.

File: src/bootstrap/builder.rs (hunk 8 above)
Comment (jyn514) [reply src/bootstrap/builder.rs:758]: I would rather inline it, yeah - there are a *lot* of tools and I don't think they all need their own function.

File: src/test/rustdoc-json/nested.rs (hunk 10 above)
Comment (jyn514) [reply src/test/rustdoc-json/nested.rs:10]: I mean hard-coding the DefIds anywhere, so both. `@count` sounds good to me.

File: src/tools/jsondocck/src/config.rs (hunk 12 above)
Comment (jyn514) [reply src/tools/jsondocck/src/config.rs:30]: I would prefer to only check it in one place, yeah.

File: src/tools/jsondocck/src/cache.rs (hunk 13 above)
Comment (jyn514) [reply src/tools/jsondocck/src/cache.rs:49]: That would be great, thanks!

File: src/bootstrap/tool.rs (hunk 9 above)
Comment (CraftSpider) [reply src/bootstrap/tool.rs:370]: Trying to remove `is_unstable_tool` triggers an assertion failure:
`thread 'main' panicked at 'assertion failed: !use_snapshot || stage == 0 || self.local_rebuild', src\bootstrap\builder.rs:1003:9`

Comment (jyn514) [reply src/bootstrap/tool.rs:370]: That's super weird, ToolBootstrap tools should always be compiled with stage 0. Can you post the backtrace?

Comment (CraftSpider) [reply src/bootstrap/tool.rs:370]: <details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'main' panicked at 'assertion failed: !use_snapshot || stage == 0 || self
.local_rebuild', src\bootstrap\builder.rs:1008:9
stack backtrace:
   0:     0x7ff61c3b9285 - std::backtrace_rs::backtrace::dbghelp::trace
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\..\..\backtrace\src\backtrace\dbghelp.rs:98
   1:     0x7ff61c3b9285 - std::backtrace_rs::backtrace::trace_unsynchronized
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\..\..\backtrace\src\backtrace\mod.rs:66
   2:     0x7ff61c3b9285 - std::sys_common::backtrace::_print_fmt
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:67
   3:     0x7ff61c3b9285 - std::sys_common::backtrace::_print::{{impl}}::fmt
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:46
   4:     0x7ff61c3d9e9b - core::fmt::write
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\core\src\fmt\mod.rs:1078
   5:     0x7ff61c3b2afd - std::io::Write::write_fmt<std::sys::windows::stdio::S
tderr>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\io\mod.rs:1519
   6:     0x7ff61c3bcf6d - std::sys_common::backtrace::_print
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:49
   7:     0x7ff61c3bcf6d - std::sys_common::backtrace::print
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:36
   8:     0x7ff61c3bcf6d - std::panicking::default_hook::{{closure}}
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:208
   9:     0x7ff61c3bca7a - std::panicking::default_hook
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:225
  10:     0x7ff61c3bd8ae - std::panicking::rust_panic_with_hook
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:591
  11:     0x7ff61c3bd3a3 - std::panicking::begin_panic_handler::{{closure}}
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:495
  12:     0x7ff61c3b9c5f - std::sys_common::backtrace::__rust_end_short_backtrac
e<closure-0,!>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\sys_common\backtrace.rs:141
  13:     0x7ff61c3bd329 - std::panicking::begin_panic_handler
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:493
  14:     0x7ff61c3d86e0 - core::panicking::panic_fmt
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\core\src\panicking.rs:92
  15:     0x7ff61c3d862c - core::panicking::panic
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\core\src\panicking.rs:50
  16:     0x7ff61baeabb9 - bootstrap::builder::Builder::cargo
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1008
  17:     0x7ff61befb922 - bootstrap::tool::prepare_tool_cargo
                               at D:\Dev\Git\rust\src\bootstrap\tool.rs:230
  18:     0x7ff61bef9cf9 - bootstrap::tool::{{impl}}::run
                               at D:\Dev\Git\rust\src\bootstrap\tool.rs:61
  19:     0x7ff61bddda62 - bootstrap::builder::Builder::ensure<bootstrap::tool::
ToolBuild>
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1494
  20:     0x7ff61bf01751 - bootstrap::tool::{{impl}}::run
                               at D:\Dev\Git\rust\src\bootstrap\tool.rs:326
  21:     0x7ff61be4b1c7 - bootstrap::builder::Builder::ensure<bootstrap::tool::
JsonDocCk>
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1494
  22:     0x7ff61bee4c1e - bootstrap::test::{{impl}}::run
                               at D:\Dev\Git\rust\src\bootstrap\test.rs:1075
  23:     0x7ff61bdbedb1 - bootstrap::builder::Builder::ensure<bootstrap::test::
Compiletest>
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1494
  24:  24:     0x7ff61bef3ad1 - bootstrap::test::{{impl}}::run
                               at D:\Dev\Git\rust\src\bootstrap\test.rs:852
  25:     0x7ff61be40575 - bootstrap::builder::Builder::ensure<bootstrap::test::
RustdocJson>
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:1494
  26:     0x7ff61bef3957 - bootstrap::test::{{impl}}::make_run
                               at D:\Dev\Git\rust\src\bootstrap\test.rs:848
  27:     0x7ff61bae30cc - bootstrap::builder::StepDescription::maybe_run
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:179
  28:     0x7ff61bae38df - bootstrap::builder::StepDescription::run
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:217
  29:     0x7ff61bae7a1f - bootstrap::builder::Builder::run_step_descriptions
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:570
  30:     0x7ff61bae78a3 - bootstrap::builder::Builder::execute_cli
                               at D:\Dev\Git\rust\src\bootstrap\builder.rs:561
  31:     0x7ff61b904334 - bootstrap::Build::build
                               at D:\Dev\Git\rust\src\bootstrap\lib.rs:506
  32:     0x7ff61b8f2083 - bootstrap::main
                               at D:\Dev\Git\rust\src\bootstrap\bin\main.rs:30
  33:     0x7ff61b8f1c0b - core::ops::function::FnOnce::call_once<fn(),tuple<>>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\core\src\ops\function.rs:227
  34:     0x7ff61b8f2acb - std::sys_common::backtrace::__rust_begin_short_backtr
ace<fn(),tuple<>>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\std\src\sys_common\backtrace.rs:125
  35:     0x7ff61b8f1ab1 - std::rt::lang_start::{{closure}}<tuple<>>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\std\src\rt.rs:66
  36:     0x7ff61c3bdba4 - core::ops::function::impls::{{impl}}::call_once
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\core\src\ops\function.rs:280
  37:     0x7ff61c3bdba4 - std::panicking::try::do_call
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:379
  38:     0x7ff61c3bdba4 - std::panicking::try
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panicking.rs:343
  39:     0x7ff61c3bdba4 - std::panic::catch_unwind
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\panic.rs:396
  40:     0x7ff61c3bdba4 - std::rt::lang_start_internal
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\/library\std\src\rt.rs:51
  41:     0x7ff61b8f1a83 - std::rt::lang_start<tuple<>>
                               at /rustc/05b6023675d77979637b04a350c85903fbf5925
7\library\std\src\rt.rs:65
  42:     0x7ff61b8f26a0 - main
  43:     0x7ff61c3e1564 - invoke_main
                               at d:\A01\_work\6\s\src\vctools\crt\vcstartup\src
\startup\exe_common.inl:78
  44:     0x7ff61c3e1564 - __scrt_common_main_seh
                               at d:\A01\_work\6\s\src\vctools\crt\vcstartup\src
\startup\exe_common.inl:288
  45:     0x7ffbb4fb7034 - BaseThreadInitThunk
  46:     0x7ffbb6edd0d1 - RtlUserThreadStart
failed to run: D:\Dev\Git\rust\build\bootstrap\debug\bootstrap test src/test/rus
tdoc-json
Build completed unsuccessfully in 0:00:38

Process finished with exit code 1
```

</p>
</details>

Comment (CraftSpider) [reply src/bootstrap/tool.rs:370]: It looks like it fails because it's not a rebuild, it's during stage 1, and if *not* unstable, it *is* ToolBootstrap, which is apparently not allowed in conjunction with the other two? My first guess is that it relates to compiletest being unstable, but that's not based on any evidence.

File: src/bootstrap/test.rs [hunk 16]
```diff
@@ -1072,6 +1072,8 @@ note: if you're sure you want to do this, please open an issue as to why. In the
 
         cmd.arg("--docck-python").arg(builder.python());
 
+        cmd.arg("--jsondocck-path").arg(builder.ensure(tool::JsonDocCk { compiler, target }));
```
Comment (jyn514) [inline src/bootstrap/test.rs:1075]: I found the issue: `compiler` here should always be the bootstrap compiler. You're trying to compile it with stage 1. Try using stage 0 instead.

Comment (CraftSpider) [reply src/bootstrap/test.rs:1075]: How would I get the stage0 compiler here? I see above it seems to use the same compiler for rustdoc, so I'm not sure how to get the older one.

Comment (CraftSpider) [reply src/bootstrap/test.rs]: I made it use stage0, I also made jsondocck only build when testing rustdoc-json to match rustdoc's behavior in compiletest

**Comment by jyn514** [comment]
@bors r+

Thanks so much!
